lib/galaxy/tool_util/linters/stdio.py

```diff
@@ -1,68 +1,78 @@
+                        node=tool_node,
+                    )
+
+
+class StdIOMultiple(Linter):
+    code = "S002"
+
+    def lint(tool_source: "ToolSource", lint_ctx: "LintContext"):
+        tool_xml = getattr(tool_source, "xml_tree", None)
+        if not tool_xml:
+            # Can only lint XML tools at this point.
+            # Should probably use tool_source.parse_stdio() to abstract away XML details
+            return
+        stdios = tool_xml.findall("./stdio") if tool_xml else []
 
-    if len(stdios) > 1:
-        lint_ctx.error("More than one stdio tag found, behavior undefined.", node=stdios[1])
-        return
+        if len(stdios) > 1:
+            lint_ctx.error("More than one stdio tag found, behavior undefined.", node=stdios[1])
+            return
 
-    stdio = stdios[0]
-    for child in list(stdio):
-        if child.tag is etree.Comment:
-            continue
-        if child.tag == "regex":
-            _lint_regex(tool_xml, child, lint_ctx)
-        elif child.tag == "exit_code":
-            _lint_exit_code(tool_xml, child, lint_ctx)
-        else:
-            message = "Unknown stdio child tag discovered [%s]. "
-            message += "Valid options are exit_code and regex."
-            lint_ctx.warn(message % child.tag, node=child)
 
+class StdIORegex(Linter):
+    code = "S003"
 
-def _lint_exit_code(tool_xml, child, lint_ctx):
-    for key in child.attrib.keys():
-        if key not in ["description", "level", "range"]:
-            lint_ctx.warn(f"Unknown attribute [{key}] encountered on exit_code tag.", node=child)
+    def lint(tool_source: "ToolSource", lint_ctx: "LintContext"):
+        tool_xml = getattr(tool_source, "xml_tree", None)
+        if not tool_xml:
+            # Can only lint XML tools at this point.
+            # Should probably use tool_source.parse_stdio() to abstract away XML details
+            return
+        stdios = tool_xml.findall("./stdio") if tool_xml else []
 
+        if len(stdios) != 1:
+            return
 
-def _lint_regex(tool_xml, child, lint_ctx):
-    for key in child.attrib.keys():
-        if key not in ["description", "level", "match", "source"]:
-            lint_ctx.warn(f"Unknown attribute [{key}] encountered on regex tag.", node=child)
-    match = child.attrib.get("match")
-    if match:
-        try:
-            re.compile(match)
-        except Exception as e:
-            lint_ctx.error(f"Match '{match}' is no valid regular expression: {str(e)}", node=child)
+        stdio = stdios[0]
+        for child in list(stdio):
+            if child.tag == "regex":
+                match = child.attrib.get("match")
+                if match:
+                    try:
+                        re.compile(match)
+                    except Exception as e:
+                        lint_ctx.error(f"Match '{match}' is no valid regular expression: {str(e)}", node=child)
```

Comment:
Don't we need to have a code per message, unless one linter only emits one possible message ?

Reply:
We can try to do this. With the current state I would need to split `StdIOAbsence` only, or?  I hoped to come away with something like at most one message per linter (which would be fine for profile versions as in `StdIOAbsence`.

Reply:
Also, it's important to note that we currently filter for linter function/class names ([here](https://github.com/galaxyproject/galaxy/blob/00c4fd05ca668bb9685e40ab83a1bc11e837c945/lib/galaxy/tool_util/lint.py#L204)) and not linter message codes. I would say that do not need the codes. But we can either name the filter classes using a code.

Reply:
One message per linter sounds fine, but also like a lot of work.

Reply:
> One message per linter sounds fine, but also like a lot of work.

The good thing is that we have good unit test coverage - which now pays off.

Reply:
If you augment the lint_ctx.function to instead pass codes we can use those codes to determine what is classified as an error / what can be ignored. This seems like less work (we'd have to run all relatively coarse linters, but I think that's ok ?)

Reply:
Excellent idea, but I think we are fixed to the signature of the linter function, i.e 1st parameter `ToolSource` and 2nd parameter `LintContext`. The reason is that there are some linters that live in the planemo sources. One option would be to move them here (not sure if this would add extra requirements) - which might be good anyway.

But there might be another way, due to this [hack](https://github.com/galaxyproject/galaxy/blob/00c4fd05ca668bb9685e40ab83a1bc11e837c945/lib/galaxy/tool_util/lint.py#L213) we can filter the messages before adding them to the final message list (and also if without this hack we could filter the list of messages after each linter application).

I'm a bit worried how we can ensure that our codes are not duplicated. 

> One message per linter sounds fine, but also like a lot of work.

Main question for me is what is the "best" solution .. 

> we'd have to run all relatively coarse linters, but I think that's ok ?

I also think this would be OK. Runtime should not be a concern.

Reply:
>  The reason is that there are some linters that live in the planemo sources

I think when I first implemented the XSD validation the external dependency on xmllint (https://github.com/galaxyproject/planemo/blob/master/planemo/xml/validation.py#L14C20-L14C27) was the issue. lxml seemed less mature at the time than xmllint. I guess we use lxml a lot more aggressively now and we have been generally pleased with the XSD validation it is doing and we now have a Galaxy dependency on lxml. I think I would be fine dropping the xmllint path through planemo if it would simplify things and let the XSD validation move into Galaxy.

Reply:
Thanks for the comment @jmchilton. Will try to move it here, but probably in a separate PR. 

I'm still experimenting if separate classes are the way to go. It seems difficult to avoid code duplication, but it will become clearer how we end up in the different linter messages.